src/util/hd-xinput.c

```diff
@@ -6,20 +6,27 @@
 #include <X11/extensions/Xrandr.h>
 #include <clutter/clutter-main.h>
 #include <clutter/x11/clutter-x11.h>
+#include <matchbox/core/mb-wm.h>
+#include "home/hd-render-manager.h"
 
 #define RR_Reflect_All	(RR_Reflect_X|RR_Reflect_Y)
 
 GArray *xi_devices = NULL;
 int xi_motion_ev_type = -1;
 int xi_presence_ev_type = -1;
 
+typedef struct {
+	gboolean is_ts;
+	XDevice *dev;
+} hd_xi_device;
+
 typedef struct Matrix {
 	float m[9];
 } Matrix;
 
 void hd_close_input_devices(Display * dpy)
 {
-	for (int i = 0; i < xi_devices->len; i++) {
+	for (int i = 0; xi_devices && i < xi_devices->len; i++) {
```

Comment:
are you sure the compiler will optimize that?

Reply:
yes ofc: https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:20,endLineNumber:8,positionColumn:20,positionLineNumber:8,selectionStartColumn:20,selectionStartLineNumber:8,startColumn:20,startLineNumber:8),source:'//+Type+your+code+here,+or+load+an+example.%0Aint+square(int+*ptr,+int+*num)+%7B%0A++++for(int+i+%3D+0%3B+ptr+%26%26+i+%3C+10%3B+i%2B%2B)+%7B%0A++++++++(*num)%2B%3D*ptr%3B%0A++++%7D%0A++++return+0%3B%0A%7D%0Aint+circle(int+*ptr,+int+*num)+%7B%0A++++if(ptr)+%7B%0A++++++++for(int+i+%3D+0%3B+i+%3C+10%3B+i%2B%2B)+%7B%0A++++++++++++(*num)%2B%3D*ptr%3B%0A++++++++%7D%0A++++%7D%0A++++return+0%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:arm1030,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'-O2+-shared',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1,tree:'1'),l:'5',n:'0',o:'ARM+gcc+10.3+(linux)+(C%2B%2B,+Editor+%231,+Compiler+%231)',t:'0')),k:33.333333333333336,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compiler:1,editor:1,fontScale:14,fontUsePx:'0',tree:'1',wrap:'1'),l:'5',n:'0',o:'Output+of+ARM+gcc+10.3+(linux)+(Compiler+%231)',t:'0')),k:33.33333333333333,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4